### PR TITLE
Tock Tutorial 2023

### PR DESCRIPTION
**Pull request by lschuermann**
USB Security Key writeup!

Relies on tock/libtock-c#335 and tock/tock#3562 for the application and kernel side of things.

For the access control stuff, it also relies on:
 * Tockloader release
 * elf2tab release
 * tock/tock#3557 (probably)
 * tock/libtock-c#327

## Outstanding TODOs

1. Encryption Oracle Capsule. Need to figure out testing.
2. ~~Access Control. Need to figure out `pconsole` stuff~~

### CODE REVIEW

File: src/key-hotp-oracle.md [hunk 1]
```diff
@@ -0,0 +1,865 @@
...
+use kernel::hil::symmetric_encryption::Client;
+
+impl<'a, A: AES128<'a> + AES128Ctr> Client<'a> for EncryptionOracleDriver<'a, A> {
+    fn crypt_done(&'a self, mut source: Option<&'static mut [u8]>, destination: &'static mut [u8]) {
+	    unimplemented!()
+    }
+}
+```
+
+With this trait implemented, we can wire up the `oracle` driver instance to
```
Comment (alexandruradovici) [inline src/key-hotp-oracle.md:558]: Explain what happens if this line is not present, this is a common mistake.

File: src/key-hotp-oracle.md [hunk 2]
```diff
@@ -0,0 +1,865 @@
...
+```diff
+  use kernel::utilities::cells::OptionalCell;
+
+  pub struct EncryptionOracleDriver<'a, A: AES128<'a> + AES128Ctr> {
+      aes: &'a A,
+      process_grants: Grant<ProcessState, UpcallCount<0>, AllowRoCount<0>, AllowRwCount<0>>,
++     current_process: OptionalCell<ProcessId>,
+  }
+```
+
```
Comment (alexandruradovici) [inline src/key-hotp-oracle.md:594]:
```suggestion
```diff
pub fn new(
        aes: &'a A,
        _source_buffer: &'static mut [u8],
        _dest_buffer: &'static mut [u8],
        process_grants: Grant<ProcessState, UpcallCount<0>, AllowRoCount<0>, AllowRwCount<0>>,
    ) -> Self {
        EncryptionOracleDriver {
            process_grants,
            aes,
+            current_process: OptionalCell::empty(),
        }
    }
```
```

Make sure you add the closing ``` here, github seems to use the ``` to as ending for the suggestion.

File: src/key-hotp-oracle.md [hunk 3]
```diff
@@ -0,0 +1,865 @@
...
+other code (or potentially the application itself) is executing.
+
+For this reason, Tock uses _static mutable slices_ (`&'static mut [u8]`) in
+HILs. These Rust types have the distinct advantage that they can be passed
+around the kernel as "persistent references": when borrowing a `'static`
+reference into another `'static` reference, the original reference becomes
+inaccessible. Tock features a special container to hold such mutable references,
+called `TakeCell`. We add such a container for each of our source and
+destination buffers:
+
```
Comment (alexandruradovici) [inline src/key-hotp-oracle.md:689]: Import `TakeCell` and import `Cell`.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: src/key-hotp-oracle.md [hunk 4]
```diff
@@ -0,0 +1,865 @@
...
+        aes_src_buffer,
+        aes_dst_buffer,
+		// Magic incantation to create our `Grant` struct:
+        board_kernel.create_grant(
+            0x99999, // our driver number
+            &create_capability!(capabilities::MemoryAllocationCapability)
+        ),
+    ),
+);
+
```
[the remark this answers is not in the document]
Comment (lschuermann) [reply src/key-hotp-oracle.md:429]: Done on the demo board / will be moved to `extra/src/tutorials` anyways apparently.

File: src/key-hotp-oracle.md (hunk 2 above)
Comment (lschuermann) [reply src/key-hotp-oracle.md:594]: > Make sure you add the closing ``` here, github seems to use the ``` to as ending for the suggestion.

10/10, thanks GitHub!

File: src/key-hotp-oracle.md [hunk 5]
```diff
@@ -0,0 +1,865 @@
...
+For this reason, Tock uses _static mutable slices_ (`&'static mut [u8]`) in
+HILs. These Rust types have the distinct advantage that they can be passed
+around the kernel as "persistent references": when borrowing a `'static`
+reference into another `'static` reference, the original reference becomes
+inaccessible. Tock features a special container to hold such mutable references,
+called `TakeCell`. We add such a container for each of our source and
+destination buffers:
+
+```diff
+  pub struct EncryptionOracleDriver<'a, A: AES128<'a> + AES128Ctr> {
```
Comment (lschuermann) [inline src/key-hotp-oracle.md:691]:
```suggestion
  use core::cell::Cell;
  use kernel::utilities::cells::TakeCell;

  pub struct EncryptionOracleDriver<'a, A: AES128<'a> + AES128Ctr> {
```

File: src/key-hotp-oracle.md [hunk 6]
```diff
@@ -0,0 +1,865 @@
...
+>         Finished release [optimized + debuginfo] target(s) in 20.34s
+>        text    data     bss     dec     hex filename
+>      176132       4   33284  209420   3320c /home/tock/tock/target/thumbv7em-none-eabi/release/nrf52840dk
+>     [Hash ommitted]  /home/tock/tock/target/thumbv7em-none-eabi/release/nrf52840dk.bin
+
+Applications interact with our driver by passing a "driver number" alongside
+their system calls. The `capsules/core/src/driver.rs` module acts as a registry
+for driver numbers. For the purposes of this tutorial we'll use an unassigned
+driver number in the _misc_ range, `0x99999`.
+
```
[the remark this answers is not in the document]
Comment (lschuermann) [reply src/key-hotp-oracle.md:347]: Added (but not to `driver.rs`) and PRed: https://github.com/tock/tock/pull/3567

**Pull request merged**